numpy/core/src/umath/ufunc_object.c

```diff
@@ -5105,6 +5223,28 @@ PyUFunc_FromFuncAndDataAndSignatureAndIdentity(PyUFuncGenericFunction *func, voi
             return NULL;
         }
     }
+
+    char *curr_types = ufunc->types;
+    for (int i = 0; i < ntypes * (nin + nout); i += nin + nout) {
+        /*
+         * Add all legacy wrapping loops here. This is normally necessary, but
+         * makes sense.  It could also help/be needed to avoid issues with
+         * ambiguous loops such as: `OO->?` and `OO->O` where in theory the
+         * wrong loop could be picked if only the second was added.
+         */
+        PyObject *info;
+        PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+        for (int arg = 0; arg < nin + nout; arg++) {
+            op_dtypes[arg] = PyArray_DTypeFromTypeNum(curr_types[arg]);
+            Py_DECREF(op_dtypes[arg]);  /* That DType can't be deleted... */
```

Comment:
Took a bit of time to grok the comment (here and below): I think the point is that the DType cannot be deleted in between here and where the loop is added, so we might as well decref it now rather than have a separate decref later, correct?

Reply:
Tweaked the comment to `These DTypes are immortal and adding INCREFs: so borrow it` (if still not obvious, will just explicitly decref them later).